build.py

```diff
@@ -0,0 +1,418 @@
+            f"self.packer_path",
+            f"init",
+            f"{DIR_PATH}/templates",
+        ]
+
+    def assemble_packer_build_command(self):
+        return " ".join(
+            [
+                f"{self.packer_path}",
+                f"build",
+                f"-only=qemu.{self.template}",
+                f"{DIR_PATH}/templates",
+            ]
+        )
+
+    def assemble_convert_command(self):
+        return " ".join(
+            [
+                f"{self.qemu_path}",
+                f"convert",
+                f"-O",
+                f"raw",
+                f"./images/{self.template}",
+                f"{self.image_path}",
+            ]
+        )
+
+    def assemble_os_command(self):
+        return " ".join(
+            [
+                f"{self.openstack_path}",
+                f"image",
+                f"create",
+                f"--file",
+                f"{self.image_path}",
+                f"{self.image_name}",
+            ]
+        )
+
+    def assemble_packer_envs(self):
+        env = os.environ.copy()
+        env["PACKER_PLUGIN_PATH"] = f"{DIR_PATH}/packer_plugins"
+        env[
+            "PKR_VAR_groups"
+        ] = f"[{','.join('\"' + x + '\"' for x in self.provisioning)}]"
+        env["PKR_VAR_headless"] = "true"
+        if self.ansible_args:
+            env["PKR_VAR_ansible_extra_args"] = self.ansible_args
+        return env
+
+    def assemble_name(self):
+        """
+        Uses a naming scheme described in\n
+        https://github.com/usegalaxy-eu/vgcn/issues/78
+        """
+        name = ["vgcn"]
+        if "generic" in self.provisioning:
+            provisioning = self.provisioning
+            provisioning.remove("generic")
+            name += [f"+{'+'.join(provisioning)}"]
+        else:
+            name += [f"!generic+{'+'.join(self.provisioning)}"]
+        name += [
+            "-".join(self.template.split("-", 2)[:2]),
+            self.assemble_timestamp(),
+            subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"])
+            .decode("ascii")
+            .strip(),
+            subprocess.check_output(["git", "rev-parse", "--short", "HEAD"])
+            .decode("ascii")
+            .strip(),
+        ]
+        if self.comment:
+            name += [self.comment]
+        return "~".join(name)
+
+    def assemble_scp_command(self):
+        return " ".join(
+            [
+                f"scp",
+                f"-i",
+                f"{self.pvt_key}",
+                f"{self.image_path}",
+                f"{SSH_USER}@{SSH_HOST}:{STATIC_DIR}/{os.path.basename(self.image_path)}",
+            ]
+        )
+
+    def assemble_ssh_command(self):
+        return " ".join(
+            [
+                f"ssh",
+                f"-i",
+                f"{self.pvt_key}",
+                f"{SSH_USER}@{SSH_HOST}",
+                f"chmod",
+                f"ugo+r",
+                f"{STATIC_DIR}{os.path.basename(self.image_path)}",
```

Comment:
```suggestion
                f"{STATIC_DIR / self.image_path.basename()}",
```

(only if you converted DIR_PATH to a `pathlib.Path` object)

In general you may be wondering, why? Because with with `Path` objects you do not have to care whether strings referring to folders end with `/`. An alternative is to always put a `/` between the two strings (but I think I have seen some badly implemented libraries that do not accept repeated `/`).

Reply:
It's called operator overloading, right? looks much cleaner!

Reply:
Yes, exactly!